### PR TITLE
feat(framework) Support non-`str` config value types

### PR DESCRIPTION
**Pull request by charlesbvll**
### Checklist

- [ ] Implement proposed change
- [ ] Write tests
- [ ] Update [documentation](https://flower.ai/docs/writing-documentation.html)
- [ ] Make CI checks pass
- [ ] Ping maintainers on [Slack](https://flower.ai/join-slack/) (channel `#contributions`)

### CODE REVIEW

File: src/py/flwr/cli/config_utils.py [hunk 1]
```diff
@@ -145,8 +137,6 @@ def validate_fields(config: Dict[str, Any]) -> Tuple[bool, List[str], List[str]]
     else:
         if "publisher" not in config["tool"]["flwr"]["app"]:
             errors.append('Property "publisher" missing in [tool.flwr.app]')
-        if "config" in config["tool"]["flwr"]["app"]:
-            _validate_run_config(config["tool"]["flwr"]["app"]["config"], errors)
```
Comment (danieljanes) [inline src/py/flwr/cli/config_utils.py:149]: We cannot remove this validation. We should instead update it to check if only allowed value types are configured.

File: src/py/flwr/common/config_test.py [hunk 2]
```diff
@@ -168,8 +170,8 @@ def test_get_project_config_file_valid(tmp_path: Path) -> None:
         clientapp = "fedgpt.client:app"
 
         [tool.flwr.app.config]
-        num_server_rounds = "10"
-        momentum = "0.1"
+        num_server_rounds = 10
+        momentum = 0.1
```
Comment (danieljanes) [inline src/py/flwr/common/config_test.py:174]: Should we also add a boolean here?

File: src/py/flwr/cli/config_utils.py (hunk 1 above)
Comment (charlesbvll) [reply src/py/flwr/cli/config_utils.py:149]: Good point, I'll update this

File: src/py/flwr/common/config.py [hunk 3]
```diff
@@ -112,29 +113,30 @@ def get_fused_config(run: Run, flwr_dir: Optional[Path]) -> Dict[str, str]:
...
             items.extend(flatten_dict(v, parent_key=new_key).items())
-        elif isinstance(v, str):
-            items.append((new_key, v))
+        elif isinstance(v, get_args(UserConfigValue)):
+            items.append((new_key, cast(UserConfigValue, v)))
         else:
             raise ValueError(
-                f"The value for key {k} needs to be a `str` or a `dict`.",
+                f"The value for key {k} needs to be of type `int`, `float`, "
+                "`bool, `str`, a `list` of those, or  a `dict` of those.",
```
Comment (danieljanes) [inline src/py/flwr/common/config.py:129]: We're not allowing `list` values

File: src/py/flwr/server/superlink/driver/driver_servicer.py [hunk 4]
```diff
@@ -149,7 +150,16 @@ def GetRun(
...
+        run_proto = (
+            None
+            if run is None
+            else Run(
+                run_id=run.run_id,
+                fab_id=run.fab_id,
+                fab_version=run.fab_version,
+                override_config=user_config_to_proto(run.override_config),
+            )
+        )
```
Comment (danieljanes) [inline src/py/flwr/server/superlink/driver/driver_servicer.py:162]: We could unpack this into a regular if/else for better readability

File: src/py/flwr/server/superlink/fleet/message_handler/message_handler.py [hunk 5]
```diff
@@ -113,5 +114,14 @@ def get_run(
...
+    run_proto = (
+        None
+        if run is None
+        else Run(
+            run_id=run.run_id,
+            fab_id=run.fab_id,
+            fab_version=run.fab_version,
+            override_config=user_config_to_proto(run.override_config),
+        )
+    )
```
Comment (danieljanes) [inline src/py/flwr/server/superlink/fleet/message_handler/message_handler.py:126]: Same unpacking opportunity as in `driver_servicer`

File: src/py/flwr/superexec/deployment.py [hunk 6]
```diff
@@ -84,12 +86,12 @@ def set_config(
         if not config:
             return
         if superlink_address := config.get("superlink"):
-            self.superlink = superlink_address
+            self.superlink = str(superlink_address)
```
Comment (danieljanes) [inline src/py/flwr/superexec/deployment.py:89]: Before casting, we should check if the type is actually `str` and show a meaningful error message it it's not

File: src/py/flwr/superexec/deployment.py [hunk 7]
```diff
@@ -84,12 +86,12 @@ def set_config(
...
         if root_certificates := config.get("root-certificates"):
-            self.root_certificates = root_certificates
-            self.root_certificates_bytes = Path(root_certificates).read_bytes()
+            self.root_certificates = str(root_certificates)
+            self.root_certificates_bytes = Path(str(root_certificates)).read_bytes()
```
Comment (danieljanes) [inline src/py/flwr/superexec/deployment.py:92]: Same as above

File: src/py/flwr/superexec/deployment.py [hunk 8]
```diff
@@ -84,12 +86,12 @@ def set_config(
...
         if flwr_dir := config.get("flwr-dir"):
-            self.flwr_dir = flwr_dir
+            self.flwr_dir = str(flwr_dir)
```
Comment (danieljanes) [inline src/py/flwr/superexec/deployment.py:94]: Same as above

File: src/py/flwr/superexec/simulation.py [hunk 9]
```diff
@@ -42,43 +43,45 @@ class SimulationEngine(Executor):
...
-            - "num-supernodes": str
+            - "num-supernodes": int
                 Number of nodes to register for the simulation.
         """
         if not config:
             return
-        if num_supernodes := config.get("num-supernodes"):
+        num_supernodes = config.get("num-supernodes")
+        if num_supernodes and isinstance(num_supernodes, int):
             self.num_supernodes = num_supernodes
```
Comment (danieljanes) [inline src/py/flwr/superexec/simulation.py:69]: Should we perhaps display a meaningful error message here if the type is wrong?

File: src/py/flwr/cli/config_utils.py [hunk 10]
```diff
@@ -112,8 +113,11 @@ def _validate_run_config(config_dict: Dict[str, Any], errors: List[str]) -> None
     for key, value in config_dict.items():
         if isinstance(value, dict):
             _validate_run_config(config_dict[key], errors)
-        elif not isinstance(value, str):
-            errors.append(f"Config value of key {key} is not of type `str`.")
+        elif not isinstance(value, get_args(UserConfigValue)):
+            raise ValueError(
+                f"The value for key {key} needs to be of type `int`, `float`, "
+                "`bool, `str`, or  a `dict` of those.",
```
Comment (danieljanes) [inline src/py/flwr/cli/config_utils.py:119]: Wouldn't `dict` be an error at this point?
```suggestion
                f"The value for key {key} needs to be of type `bool`, `float`, `int` "
                "or `str`.",
```

File: src/py/flwr/common/config.py [hunk 11]
```diff
@@ -112,29 +113,30 @@ def get_fused_config(run: Run, flwr_dir: Optional[Path]) -> Dict[str, str]:
...
             items.extend(flatten_dict(v, parent_key=new_key).items())
-        elif isinstance(v, str):
-            items.append((new_key, v))
+        elif isinstance(v, get_args(UserConfigValue)):
+            items.append((new_key, cast(UserConfigValue, v)))
         else:
             raise ValueError(
-                f"The value for key {k} needs to be a `str` or a `dict`.",
+                f"The value for key {k} needs to be of type `int`, `float`, "
+                "`bool, `str`, or  a `dict` of those.",
```
Comment (danieljanes) [inline src/py/flwr/common/config.py:129]: Same question as above

File: src/py/flwr/superexec/deployment.py [hunk 12]
```diff
@@ -84,12 +86,18 @@ def set_config(
         if not config:
             return
         if superlink_address := config.get("superlink"):
+            if not isinstance(superlink_address, str):
+                raise ValueError("The `superlink` value should be a string.")
```
Comment (danieljanes) [inline src/py/flwr/superexec/deployment.py:90]:
```suggestion
                raise ValueError("The `superlink` value should be of type `str`.")
```

File: src/py/flwr/superexec/deployment.py [hunk 13]
```diff
@@ -84,12 +86,18 @@ def set_config(
         if not config:
             return
         if superlink_address := config.get("superlink"):
+            if not isinstance(superlink_address, str):
+                raise ValueError("The `superlink` value should be of type `str`.")
             self.superlink = superlink_address
         if root_certificates := config.get("root-certificates"):
+            if not isinstance(root_certificates, str):
+                raise ValueError("The `root-certificates` value should be a string.")
```
Comment (danieljanes) [inline src/py/flwr/superexec/deployment.py:94]:
```suggestion
                raise ValueError("The `root-certificates` value should be of type `str`.")
```

File: src/py/flwr/superexec/deployment.py [hunk 14]
```diff
@@ -84,12 +86,18 @@ def set_config(
...
         if root_certificates := config.get("root-certificates"):
+            if not isinstance(root_certificates, str):
+                raise ValueError("The `root-certificates` value should be of type `str`.")
             self.root_certificates = root_certificates
-            self.root_certificates_bytes = Path(root_certificates).read_bytes()
+            self.root_certificates_bytes = Path(str(root_certificates)).read_bytes()
         if flwr_dir := config.get("flwr-dir"):
-            self.flwr_dir = flwr_dir
+            if not isinstance(flwr_dir, str):
+                raise ValueError("The `flwr-dir` value should be a string.")
```
Comment (danieljanes) [inline src/py/flwr/superexec/deployment.py:99]:
```suggestion
                raise ValueError("The `flwr-dir` value should be of type `str`.")
```

File: src/py/flwr/superexec/simulation.py [hunk 15]
```diff
@@ -42,43 +43,46 @@ class SimulationEngine(Executor):
...
             Supported configuration key/value pairs:
-            - "num-supernodes": str
+            - "num-supernodes": int
                 Number of nodes to register for the simulation.
         """
         if not config:
             return
         if num_supernodes := config.get("num-supernodes"):
+            if not isinstance(num_supernodes, int):
+                raise ValueError("The `num-supernodes` value should be an `int`.")
```
Comment (danieljanes) [inline src/py/flwr/superexec/simulation.py:69]:
```suggestion
                raise ValueError("The `num-supernodes` value should be of type `int`.")
```

**Review by danieljanes — DISMISSED** [review]

**Review by danieljanes — APPROVED** [review]

**Pull request merged**